File: facade/src/main/java/org/tdf/sunflower/state/Address.java
package org.tdf.sunflower.state;

import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import lombok.NonNull;
import org.tdf.common.crypto.ECKey;
import org.tdf.common.util.HexBytes;
import org.tdf.sunflower.types.Transaction;


@NoArgsConstructor(access = AccessLevel.PRIVATE)
public final class Address {
    static HexBytes EMPTY = HexBytes.fromBytes(new byte[Transaction.ADDRESS_LENGTH]);

    public static HexBytes empty() {
        return EMPTY;
    }


    public static HexBytes of(@NonNull String hex) {
        HexBytes ret = HexBytes.fromHex(hex);
        if (ret.size() == Transaction.ADDRESS_LENGTH) return ret;
        throw new RuntimeException("invalid hex, not an address");
    }

    public static HexBytes fromPrivate(HexBytes privateK){
        return fromPrivate(privateK.getBytes());
    }

    public static HexBytes fromPrivate(byte[] privateK){
        return HexBytes.fromBytes(ECKey.fromPrivate(privateK).getAddress());
    }

    public static HexBytes fromPub(HexBytes pub) {
        return HexBytes.fromBytes(ECKey.fromPublicOnly(pub.getBytes()).getAddress());
    }
}

File: facade/src/main/java/org/tdf/sunflower/state/Address.kt
package org.tdf.sunflower.state

import org.tdf.common.crypto.ECKey
import org.tdf.common.util.HexBytes
import org.tdf.sunflower.types.Transaction

object Address {
    val EMPTY: HexBytes = HexBytes.fromBytes(ByteArray(Transaction.ADDRESS_LENGTH))

    @JvmStatic
    fun empty(): HexBytes {
        return EMPTY
    }

    @JvmStatic
    fun of(hex: String): HexBytes {
        val ret = hex.hex()
        if (ret.size() == Transaction.ADDRESS_LENGTH) return ret
        throw RuntimeException("invalid hex, not an address")
    }

    @JvmStatic
    fun fromPrivate(privateK: ByteArray): HexBytes {
        return ECKey.fromPrivate(privateK).address.hex()
    }
}